libraries/AP_OpenDroneID/AP_OpenDroneID.cpp

```diff
@@ -0,0 +1,607 @@
+    // @Values: -1:Disabled,0:Serial0,1:Serial1,2:Serial2,3:Serial3,4:Serial4,5:Serial5,6:Serial6
+    AP_GROUPINFO("MAVPORT", 2, AP_OpenDroneID, _mav_port, -1),
+
+    // @Param: CANDRIVER
+    // @DisplayName: DroneCAN driver number
+    // @Description: DroneCAN driver index, 0 to disable DroneCAN
+    // @Values: 0:Disabled,1:Driver1,2:Driver2
+    AP_GROUPINFO("CANDRIVER", 3, AP_OpenDroneID, _can_driver, 0),
+    
+    // @Param: OPTIONS
+    // @DisplayName: OpenDroneID options
+    // @Description: Options for OpenDroneID subsystem. Bit 0 means to enforce arming checks
+    // @Bitmask: 0:EnforceArming
+    AP_GROUPINFO("OPTIONS", 4, AP_OpenDroneID, _options, 0),
+
+    // @Param: BARO_ACC
+    // @DisplayName: Barometer vertical accuraacy
+    // @Description: Barometer Vertical Accuracy when installed in the vehicle. Note this is dependent upon installation conditions and thus disabled by default
+    // @Units: meters
+    // @User: Advanced
+    AP_GROUPINFO("BARO_ACC", 5, AP_OpenDroneID, _baro_accuracy, -1.0),
+
+    AP_GROUPEND
+};
+
+// constructor
+AP_OpenDroneID::AP_OpenDroneID()
+{
+#if CONFIG_HAL_BOARD == HAL_BOARD_SITL
+    if (_singleton != nullptr) {
+        AP_HAL::panic("OpenDroneID must be singleton");
+    }
+#endif
+    _singleton = this;
+    AP_Param::setup_object_defaults(this, var_info);
+}
+
+void AP_OpenDroneID::init()
+{
+    if (_enable == 0) {
+        return;
+    }
+
+    _chan = mavlink_channel_t(gcs().get_channel_from_port_number(_mav_port));
+    _mavlink_dynamic_period_ms = 1000; // send dynamic messages at 1 Hz frequency i.e. every 1 second
+    _mavlink_static_period_ms = 10000; // send dynamic messages at 0.1 Hz frequency i.e. every 10 seconds
+}
+
+// Perform the pre-arm checks and prevent arming if they are not satisifed
+// Except in the case of an in-flight reboot
+bool AP_OpenDroneID::pre_arm_check(char* failmsg, uint8_t failmsg_len)
+{
+    WITH_SEMAPHORE(_sem);
+
+    if (!option_set(Options::EnforceArming)) {
+        return true;
+    }
+
+    if (pkt_basic_id.id_type == MAV_ODID_ID_TYPE_NONE) {
+        strncpy(failmsg, "UA_TYPE required in BasicID", failmsg_len);
+        return false;
+    }
+
+    if (pkt_system.operator_latitude == 0 && pkt_system.operator_longitude == 0) {
+        strncpy(failmsg, "operator location must be set", failmsg_len);
+        return false;
+    }
+
+    const uint32_t max_age_ms = 3000;
+    const uint32_t now_ms = AP_HAL::millis();
+
+    if (last_arm_status_ms == 0 || now_ms - last_arm_status_ms > max_age_ms) {
+        strncpy(failmsg, "ARM_STATUS not available", failmsg_len);
+        return false;
+    }
+
+    if (last_system_ms == 0 || now_ms - last_system_ms > max_age_ms) {
+        strncpy(failmsg, "SYSTEM not available", failmsg_len);
+        return false;
+    }
+    
+    if (arm_status.status != MAV_ODID_GOOD_TO_ARM) {
+        strncpy(failmsg, arm_status.error, failmsg_len);
+        return false;
+    }
+    
```

Comment:
We should add an arming check here the 

```suggestion
        if(!AP:gps().have_height_above_wgs84()) {
                strncpy(failmsg, "GPS must provide WGS84 altitude", failmsg_len);
                return false;
        }
```

Reply:
switched to undulation